### ISSUE TITLE
Having ybioath gui open stops gpg2 and gpg-agent from working

### ISSUE THREAD

**Issue by zephyia** [Yubico/yubioath-desktop#89]
when the ybioath gui is open on OSX (even in the system tray at the top) gpg-agent is unable to communicate with the yubikey, as soon as you close the program it all functions again - having the two operations being mutually exclusive, especially when using gpg-agent as an ssh-agent as well - is not an ideal scenario

**Comment by jas4711** [comment]
I believe scdaemon open the smartcard in exclusive mode.  I suspect yubioath does not require this.  Thus, this problem is induced by gpg-agent's design.  I believe that GnuPG people suggest apps to talk to smartcards through gpg-agent/scdaemon instead, or to kill scdaemon.  So I believe there are two possible positions to take here: 1) use the "kill scdaemon on show" setting in yubioath (maybe this should be the default?), or 2) implement support for detecting that gpg-agent/scdaemon is available and talk to smartcards through it.

**Comment by jas4711** [comment]
This happens on Ubuntu 16.04 too, btw.

**Comment by dagheyman** [comment]
Is this still the case with the new version (4.0.0) ?
https://developers.yubico.com/yubioath-desktop/Releases/

**Comment by dagheyman** [comment]
Closing, successfully used gpg with YubiKey, with Yubico Authenticator 4.0.0 open on macOS Sierra.

**Comment by tomgillett** [comment]
@dagheyman I'm still experiencing this issue on macOS, running Yubico Authenticator 4.1.2. Running `gpgconf --reload scdaemon` temporarily resolves the issue.

To reproduce I simply create an SSH connection (utilising gpg-agent) and open Yubikey Authenticator. Authenticator freezes on "Connecting to Yubikey". 

Are you able to provide any information on the configuration that worked for you?

**Comment by dagheyman** [comment]
@tomgillett We have [some code](https://github.com/Yubico/yubikey-manager/blob/master/ykman/driver_ccid.py#L167) trying to kill scdaemon when needed, however in a sandboxed app it seems to result in `Operation not permitted`. I'll open a new issue in the yubikey-manager project for this.

Your workaround is probably as good as it gets right now, re-inserting the YubiKey seems to work  as well.

**Comment by tomgillett** [comment]
Thanks @dagheyman. 

To confirm, running `ykman oath info` also "unblocks" Yubikey Authenticator. 

Might Yubikey Authenticator adopt the same approach to Yubikey Manager? (Presumably killing scdaemon as required). I understand this used to be an option in the Yubikey Authenticator app, but removed for 4.x?

**Comment by dagheyman** [comment]
@tomgillett It actually uses the exact same code (yubikey-manager is used as a library), the difference is that Yubico Authenticator is sandboxed and fails at killing scdaemon because of that.

**Comment by tomgillett** [comment]
@dagheyman thanks for confirming :)

**Comment by N4M3Z** [comment]
I've recently encountered this issue on macOS 11, one can either use gpg (keys on card) or Yubikey OATH, so indeed these operations are mutually exclusive. The GUI application seems to struggle with access to the YubiKey and one has to use the magic line `gpgconf --reload scdaemon` to temporarily fix it. However, it seems running `ykman oath code` works just as well but it kills the gpg-agent connection, so one has reinitialize it every time with `gpg --card-status`. It is not an ideal situation, but workable. My temporary workaround is to store OTP in password-store (https://github.com/tadfisher/pass-otp). Do I understand correctly that this isn't fixable since gpg-agent is the culprit here?

**Comment by fdennis** [comment]
@N4M3Z Correct. scdaemon (which is part of gpg) is holding exclusive access to the YubiKey (details here: https://dev.gnupg.org/T2440 possible workaround: setting a timeout in scdaemon.conf as described here: https://forum.yubico.com/viewtopicb85f.html?p=7513).

**Comment by N4M3Z** [comment]
I'm honestly surprised this isn't explicitly covered in Yubi official documentation regarding OATH. When you encounter the issue, the first response is to assume your key or your USB port/connection is broken. Any mention of this is hard to find, unless you specifically search for something along the lines of "gpg-agent oath collision". You won't find it in context of `ykman` or `password-store` either. This is a deal breaker for people who use the tokens for GPG key escrow through CCID as it essentially makes the OATH functionality unusable. I had no luck with `scdaemon.conf` on Mac OS, I don't think `scdaemon` works the same way with Mac GPG Tools and `tokend`, but I might be mistaken.
